rest/doc_api.go

```diff
@@ -221,16 +222,24 @@ func (h *handler) handlePutAttachment() error {
 		return err
 	}
 
-	body, err := h.db.GetRev(docid, revid, false, nil)
-	if err != nil && base.IsDocNotFoundError(err) {
-		// couchdb creates empty body on attachment PUT
-		// for non-existant doc id
-		body = db.Body{}
-		body[db.BodyRev] = revid
-	} else if err != nil {
-		return err
-	} else if body != nil {
-		body[db.BodyRev] = revid
+	var body db.Body
+	if rev, err := h.db.GetRev(docid, revid, false, nil); err != nil {
+		if base.IsDocNotFoundError(err) {
+			// couchdb creates empty body on attachment PUT
+			// for non-existant doc id
+			body = db.Body{db.BodyRev: revid}
+		} else if err != nil {
+			return err
+		}
+	} else if rev != nil {
+		body, err = rev.Mutable1xBody(h.db, nil, nil, false)
+		if err != nil {
+			return err
+		} else if body != nil {
+			body[db.BodyRev] = revid
```

Comment:
`revid` differs from `rev.RevID` when the given revid in the query parameter is empty. Covered by the first PUT in `TestManualAttachment`.

Reply:
Ok.  To begin with, it looks like the comment describing GetRev should be updated to indicate that it's returning the selected revision, or the current revision if the revision isn't specified.

Beyond that, though - if at this point the revid is empty and GetRev returned an active revision, I think we should be returning a 409 immediately, instead of artificially setting the revid to empty in the active revision body and relying on the Put.  It seems like a likely source of future bugs to be switching to a body the caller didn't send, and overwriting it with a different revid (even if it's blank).

This behaviour didn't show up in this PR, though, so it's a candidate for splitting into a separate ticket.

Reply:
I agree with these comments, and I've just gone and done it as part of this branch. Not a big change.